Change Certificate API to use Duration and RenewBefore

stu-gott opened the pull request:
**What this PR does / why we need it**:
This PR adapts the names of cert rotation parameters to match KubeVirt's sister projects.

**Release note**:
```release-note
Introduced Duration and RenewBefore parameters for cert rotation. Previous values are now deprecated.
```

stu-gott commented:
/cc @rmohr 

Your insight would be useful as you were the original author of the cert rotation API

zcahana commented on pkg/virt-operator/resource/apply/certificates_test.go:
```diff
@@ -0,0 +1,89 @@
+
+var _ = Describe("Certificates", func() {
+	Context("With deprecated fields defined", func() {
+		two_days := &metav1.Duration{Duration: 48 * time.Hour}
```
Not to insult your math skills ;-), but would it be somewhat simpler to have:
```go
day := &metav1.Duration{Duration: 24 * time.Hour}
twoDays := 2 * day
threeDays := 3 * day
...
```

or even:
```go
day := &metav1.Duration{Duration: 24 * time.Hour}
...
config.CA.Duration = 5 * day
```

Also note the snake_case --> camelCase.

stu-gott replied:
concur with the snake case part, but unfortunately "metav1.Duration" is not a number so we can't apply mathematical operators to it. I think you might be thinking of "time.Duration" which is a long int representing a number of milliseconds.

zcahana replied:
Right, I overlooked the fact that `metav1.Duration` is a struct type wrapping `time.Duration` (I falsely assumed it's aliasing it...).

Could still be done with a `days(n int) metav1.Duration` helper function, but I'm not sure it's still simplifying or rather complicating. Let's leave as-in.

zcahana commented on pkg/virt-operator/resource/apply/certificates_test.go:
```diff
@@ -0,0 +1,89 @@
+		It("should use CertRotateInterval if defined", func() {
+			config.Server.Duration = two_days
+
+			By("Using only CA.Duration")
```
```go
By("Using only Server.Duration")
```

zcahana commented on pkg/virt-operator/resource/apply/certificates_test.go:
```diff
@@ -0,0 +1,89 @@
+			result := getCertDuration(config)
+			Expect(result).To(Equal(two_days))
+
+			By("Defining CARotateInterval")
```
```go
By("Defining CertRotateInterval")
```

zcahana commented on pkg/virt-operator/resource/apply/certificates.go:
```diff
@@ -7,25 +7,43 @@ import (
-		return &metav1.Duration{Duration: Duration7d}
+	if config != nil {
+		// deprecated, but takes priority to provide a smooth upgrade path
+		if config.CARotateInterval != nil {
```
Should a validating webhook ensure that the old and new APIs are not used at the same time (i.e., are mutually exclusive)?

stu-gott replied:
I'm unsure of the value in this. The point of leaving the old definitions in at all is to ease upgrade burden. Enforcing use of one or the other might increase the potential to introduce upgrade complexity.

As it stands, one completely overrides the other so I'm not sure adding a validating webhook is of much value. What do you think?

zcahana replied:
I think the value is in preventing a contradicting configuration between old and new which is silently ignored. With such validation, the contradicting configuration is early discovered (at "compile time" :-D) and can be fixed.

stu-gott replied:
Upon further investigation, virt-api does not have a validating webhook for KubeVirt objects.

This makes sense. Though this codebase defines the object, it is not the owner of any objects created, so it should not be asserting control over the contents. Ensuring new values are set is HCO's responsibility.

ashleyschuett commented on pkg/virt-operator/resource/apply/certificates.go:
```diff
@@ -7,25 +7,43 @@ import (
 func getCADuration(config *k8sv1.KubeVirtSelfSignConfiguration) *metav1.Duration {
-	if config == nil || config.CARotateInterval == nil {
-		return &metav1.Duration{Duration: Duration7d}
+	if config != nil {
```
nit: we could reduce complexity a bit if we instead define `defaultDuration` and do 

```
if config == nil {
   return defaultDuration
}
...
return defaultDuration
```

we then have to return it twice so not sure :shrug:

stu-gott commented on staging/src/kubevirt.io/client-go/api/v1/types.go:
```diff
@@ -1291,9 +1291,31 @@ type KubeVirtList struct {
+
+	// Server configuration
+	// Certs are rotated and discarded
+	Server *CertConfig `json:"server,omitempty"`
```
It's true that KubeVirt only uses 3 of the 4 values that could be defined by this approach. The point of switching anything at all though was to use exactly the same parameters as HCO/other projects.

If we switch types to something else then HCO won't be able to simply map its settings in, thus kind of defeating the point.

zcahana replied:
The goal of using a consistent configuration for HCO and other projects is clear. IMO using a 3/4 subset of this configuration for KubeVirt still provides value in that it uses a consistent language and semantics. We need to weigh this vs. the possible confusion there is in an exposing a configuration that does nothing.

As for HCO, I think it's fine for HCO to *not* blindly map it's configuration to KubeVirt, but rather apply some per-operand logic in how it configures them.

Btw, do you know whether other projects have a use for `server.renewBefore`? (i.e., is it only KubeVirt that don't have a use for it?)

stu-gott replied:
Let me re-visit this. Scrutinizing the code, I think the issue is that OverlapInterval and RenewBefore don't have to be related. If we drill down in pkg/virt-operator/resource/apply/core.go, we'll find that there's a function called NextRotationDeadline which essentially sets the RenewBefore for *both* certs and ca's to 80% of the cert's lifetime.

I think the most sensible path forward is to use a default RenewBefore of 80% and then pass this down the chain and make it explicit.

zcahana replied:
> I think the most sensible path forward is to use a default RenewBefore of 80% and then pass this down the chain and make it explicit.

Makes sense to me.

stu-gott replied:
Fixed

stu-gott commented on staging/src/kubevirt.io/client-go/api/v1/types.go:
```diff
@@ -1291,9 +1291,31 @@ type KubeVirtList struct {
+// +k8s:openapi-gen=true
+type CertConfig struct {
+	// The requested 'duration' (i.e. lifetime) of the Certificate.
+	Duration *metav1.Duration `json:"duration,omitempty"`
```
Our use of "+optional" seems to be sporadic at best. I sort of assumed it was implied by "omitempty"?

zcahana replied:
It's somewhat implied by `omitempty` (for those looking at the code), although it tends to be mindlessly added sometimes.
For my understanding, it merely affects marshaling, whereas `+optional` affects validation and is also visible via `kubectl explain`.

zcahana replied:
I've just tried and `+optional` doesn't seem to be visible via `kubectl explain`. Some field are marked as `-required-`, but I'm not sure what's the exact mechanism behind that. I can see there are both `+optional` markers as well as `optional: "true"` and `valid: "required"` struct tags - and these don't necessarily correspond with the `-required-` prints.  Puzzling.

rmohr replied:
The `optional` and `valid` tags are no longer considered. `+optional` should be picked up at this stage. Ensure that you run `make generate` after changing them. @xpivarc, keep me honest.

xpivarc replied:
I am not sure what `optional` and `valid` tags mean... Yes `+optional` | `kubebuilder:validation:Optional` | `inline`(json tag) | `omitempty`(json tag) will make the field `optional` if `required` is default(It is in our case).

stu-gott replied:
Forgive me, I'm stumbling over the language here. @rmohr did you mean you'd like me to add the "+optional" marking before line 1331, and re-run make generate?

rmohr replied:
I just tried to clarify that `+optional` is in principle now recognized. That was not always the case. I see in the api-spec that it is not marked as required. So you probably don't have to do anything if it should be optional.

stu-gott commented:
@zcahana and @ashleyschuett thank you so much for your reviews so far. I believe I've corrected or addressed all review outstanding feedback. Please correct me if I missed anything.

zcahana commented:
thanks @stu-gott, as far as I can tell this gets the job done right.

Not sure if @kubevirt-bot honors a non-maintainer's lgtm, but...
/lgtm

jean-edouard commented on pkg/virt-operator/resource/generate/components/secrets.go:
```diff
@@ -251,7 +251,7 @@ func NewCertSecrets(installNamespace string, operatorNamespace string) []*k8sv1.
 // current certificate should be rotated, 80% of the expiration of the
 // certificate.
-func NextRotationDeadline(cert *tls.Certificate, ca *tls.Certificate, duration *metav1.Duration) time.Time {
+func NextRotationDeadline(cert *tls.Certificate, ca *tls.Certificate, duration *metav1.Duration, renewBefore *metav1.Duration, caOverlapTime *metav1.Duration) time.Time {
```
`duration` seems to be unused, is it obsolete, should it be removed?

stu-gott replied:
:facepalm: good catch. Yes, the most correct action is to remove the parameter. thanks!

jean-edouard commented on pkg/virt-operator/resource/generate/components/secrets.go:
```diff
@@ -251,7 +251,7 @@ func NewCertSecrets(installNamespace string, operatorNamespace string) []*k8sv1.
 // nextRotationDeadline returns a value for the threshold at which the
 // current certificate should be rotated, 80% of the expiration of the
```
Is the `80%` figure still relevant?

stu-gott replied:
It's technically still the default, but agree that saying that in the function comment is confusing/not helpful..

jean-edouard commented:
/approve

enp0s3 commented on pkg/virt-operator/resource/generate/components/secrets.go:
```diff
@@ -273,7 +272,21 @@ func NextRotationDeadline(cert *tls.Certificate, ca *tls.Certificate, duration *
 	notAfter := cert.Leaf.NotAfter
-	deadline := notAfter.Add(-time.Duration(float64(duration.Duration) * 0.2))
+	caDeadline := notAfter
+	deadline := cert.Leaf.NotBefore.Add(renewBefore.Duration)
```
@stu-gott I have 2 questions about this calulation : 1. Is it possible that `renewBefore` can be smaller then `cert.Leaf.NotBefore`? 2. In case the `renewBefore` is bigger then the `NotBefore`, shouldn't the deadline be `cert.Leaf.NotBefore.Add((renewBefore - cert.Leaf.NotBefore).Duration)`?

stu-gott replied:
OK. I had to dig to prove to myself what would happen if the renewBefore was nonsense. it turns out this is a valid concern. the "duration" plays into the "NotAfter" field, but "renewBefore" isn't even a concept in x509--that's entirely for our accounting/scheduling that this value exists.

Thus it makes sense to account for a renewBefore that's just too late. Sadly the most recent changeset removed the duration field. :D I propose this: I'll back that function out so that we have the duration available here. If "renewBefore" is longer than the certificate duration, I think the only sane response is to fall back to the previous behavior where the "renewBefore" value didn't exist and we just calculated "80%" of the duration. What do you think?

enp0s3 replied:
@stu-gott Sounds good to me. Another option which may be more strict is to validate these values in the validating webhook and not allow a `duration < renewBefore` situation. And maybe even set a minimum interval between these two

enp0s3 replied:
@stu-gott Another nonsense scenario can be when `cert.renewBefore > ca.RenewBefore`. That actually means that server cert renewal will occur because of the CA renewal, so `cert.renewBefore` will turn meaningless.

stu-gott replied:
Actually, this isn't a concern.

This same code path is used for both CA's and Certs--so we know the CA's renewBefore occurs before it's duration. This code already checks that the Cert must be refreshed before the CA becomes invalid. Remember, there's an overlap period where both the outgoing and the incoming CA's are honored. As long as our cert is rotated before *that* cutoff, things remain sane.

stu-gott replied:
As far as the validating webhook-there isn't one. KubeVirt resources are created by operators other than KubeVirt itself.

enp0s3 replied:
@stu-gott I see that updates can be validated after [4599](https://github.com/kubevirt/kubevirt/pull/4599) got merged.

stu-gott replied:
:facepalm: I didn't know that was there. testing a patch for this now.

stu-gott replied:
fixed!

zcahana commented:
Re-adding my lgtm after the latest update. Also /holding due to @enp0s3 question to prevent auto-merge, @stu-gott feel free to /unhold at your convenience when you deem this ready for merging.

/lgtm
/hold

rmohr commented:
/hold

Want to have a quick look tomorrow morning.

enp0s3 commented on pkg/virt-operator/resource/apply/certificates.go:
```diff
@@ -1,31 +1,96 @@
-	return config.CAOverlapInterval
+	caDuration := getCADuration(config)
+	caRenewalPeriod := getCARenewalPeriod(config)
+	return &metav1.Duration{Duration: caDuration.Duration - caRenewalPeriod.Duration}
```
@stu-gott Is there any place that makes sure that `caDuration` will always be higher than `caRenewalPeriod`?

stu-gott replied:
Yes. This exact same code is used to generate both the CA and the cert.

stu-gott commented:
Rebased to fix merge conflicts

stu-gott commented:
/retest

fgimenez commented:
/retest

stu-gott commented:
/retest

rmohr commented:
I did not get to it. But I would prefer another set of eyes on this.

@davidvossel @vladikr

vladikr reviewed (commented):
I guess I need more time to wrap my head around this...

stu-gott commented on manifests/generated/kv-resource.yaml:
```diff
@@ -44,12 +44,35 @@ spec:
+                            description: The requested 'duration' (i.e. lifetime) of the Certificate.
+                            type: string
+                          renewBefore:
+                            description: The amount of time before the currently issued certificate's "notAfter" time that we will begin to attempt to renew the certificate.
```
renewBefore and renewal period are the same thing, if I understand what you're asking. Prior to this, we simply hard coded it at 80%.

stu-gott replied:
actually. I agree. the description here is confusing/misleading--but copied from HCO.

"the amount of time after the currently issued certificate's 'notBefore' time that we will begin to attempt to renew the certificate." would make more sense, and be correct.

stu-gott replied:
/hold

@tiraboschi can you clarify things here? How would you define "renewBefore"?

tiraboschi replied:
we borrowed this from CDI that borrowed it from cert-manager:
https://cert-manager.io/docs/reference/api-docs/#cert-manager.io/v1alpha2.CertificateSpec

If you set a duration of 90 days and renewBefore at 30 days, I expect that the certificates are considered valid for 3 months and renewal will be attempted within 1 month of expiration.

stu-gott commented on manifests/generated/kv-resource.yaml:
```diff
@@ -44,12 +44,35 @@ spec:
+                      ca:
+                        description: CA configuration CA certs are kept in the CA bundle as long as they are valid
+                        properties:
+                          duration:
+                            description: The requested 'duration' (i.e. lifetime) of the Certificate.
```
Fair play. By calling it duration we're using the same name as everything else in the HCO ecosystem.

stu-gott commented:
/retest

stu-gott commented:
/hold

stu-gott commented:
/retest

stu-gott commented:
/retest

stu-gott commented:
/hold cancel

Hold was added to give reviewers a chance to look at this PR, but it's not needed at this point.

stu-gott commented:
/retest

stu-gott commented:
/retest

vladikr commented on pkg/virt-operator/resource/generate/components/secrets.go:
```diff
@@ -273,7 +273,21 @@ func NextRotationDeadline(cert *tls.Certificate, ca *tls.Certificate, duration *
+		caDeadline = ca.Leaf.NotAfter
+		if caRenewBefore != nil {
+			// Set cert rotation for the middle of the period of time when CA's overlap
+			caDeadline = notAfter.Add(-time.Duration(float64(caRenewBefore.Duration) * 0.5))
```
@stu-gott I can't wrap my head around this :/

Why do we need to do this calculation? I think I'd expect us to act normally in this case.
If the CA duration is longer than the cert duration, why can't we simply start renewing `renewBefore` time before the cert expires?

stu-gott replied:
We compare the two candidate deadlines in the next step. and we take whichever one comes first. If the CA expires before the cert does then there's going to be a period of time where the cert hasn't rotated yet, but is invalid--leading to occasional TLS errors if API calls are made.

Aiming for the middle of the renewal overlap period means:
1) the new CA will have been rolled out by then so when we renew we'll use the new CA
2) there's still some time left on the old CA so we can avoid race conditions where there's still a miniscule window of time that we could try to use the old cert right at the expiry.

stu-gott replied:
If you're asking about the math itself, renewBefore is the time before the end of a cert's lifetime, so we need to subtract this from the "notAfter" (the actual validity deadline).

As per above comment, I wanted to aim for the middle of the overlap. if we take the exact deadline there's a race condition where the old now-invalid cert could be used at the last instant.

If we renew exactly at "notAfter minus renewBefore" (in other words, the exact moment a new CA will be calculated. we used to call this overlap interval), there's a race condition where we could renew an instant too early and create yet another cert with the old CA.

I split the difference to avoid both problems.

vladikr replied:
Ok, thanks. I miss understood that this calculation is needed only when the CA expires before the cert.
Could this be that even if the CA expires after the certificate, its renewBefore/2 is long enough to be before the deadline?
Perhaps we should explicitly calculate this when the CA expires before the cert.

fgimenez commented:
/test pull-kubevirt-e2e-k8s-1.19-sig-storage

vladikr commented:
Thanks!
/lgtm

stu-gott commented:
/retest

stu-gott commented:
/test  pull-kubevirt-unit-test

failure is unrelated and has passed previously

stu-gott commented:
/retest

fgimenez commented:
/hold

Holding bc builds are failing in this PR much more frequently than in others https://prow.apps.ovirt.org/pr-history/?org=kubevirt&repo=kubevirt&pr=5241 are we sure that the errors don't come from these changes? Feel free to unhold if so, I've taken a look at pod logs in failed builds and they look ok.

stu-gott commented:
/retest

stu-gott commented:
/hold cancel

We are reviewing the errors but none of them appear related to this PR. Incidentally, this PR was rebased at changeset 778f2a9.
